zkvm/src/predicate.rs

```diff
@@ -181,75 +156,205 @@ impl Into<CompressedRistretto> for Predicate {
+    fn create_merkle_leaves(progs: &Vec<Program>, blinding_key: [u8; 32]) -> Vec<PredicateLeaf> {
+        let mut t = Transcript::new(b"ZkVM.taproot-derive-blinding");
+        let n: u64 = progs.len() as u64;
+        t.commit_u64(b"n", n);
+        t.commit_bytes(b"key", &blinding_key);
+        for prog in progs.iter() {
+            let mut buf = Vec::new();
+            prog.encode(&mut buf);
+            t.commit_bytes(b"prog", &buf);
+        }
+
+        let mut leaves = Vec::new();
+        for prog in progs.iter() {
+            let blinding = rand::thread_rng().gen::<[u8; 32]>();
+            let blinding_leaf = PredicateLeaf::Blinding(blinding);
+            let program_leaf = PredicateLeaf::Program(ProgramWitness::Program(prog.clone()));
+            if blinding[0] & 1 == 0 {
+                leaves.push(blinding_leaf);
+                leaves.push(program_leaf);
+            } else {
+                leaves.push(program_leaf);
+                leaves.push(blinding_leaf);
+            }
+        }
+        leaves
+    }
+}
 
-        // dummy predicates
-        let preds = vec![
-            Predicate::Opaque(gens.B.compress()),
-            Predicate::Opaque(gens.B_blinding.compress()),
-        ];
+impl CallProof {
+    pub fn serialized_length(&self) -> usize {
+        // VerificationKey is a 32-byte array
+        // MerkleNeighbor is a 32-byte array
+        32 + 4 + self.neighbors.len() * 32
+    }
 
-        let pred = Predicate::disjunction(preds.clone()).unwrap();
-        let op = pred.prove_disjunction(&preds);
-        assert!(op.verify().is_ok());
+    pub fn decode<'a>(reader: &mut SliceReader<'a>) -> Result<Self, VMError> {
+        let mut call_proof = CallProof::default();
+        let point = reader.read_point()?;
+        let ovk = VerificationKey::from_compressed(point);
+        match ovk {
+            Some(k) => call_proof.verification_key = k,
+            None => return Err(VMError::InvalidPoint),
+        }
+        let positions = reader.read_u32()?;
+        if positions == 0 {
+            return Err(VMError::FormatError);
+        }
+        let mut neighbors = vec![];
+        let num_neighbors = positions.leading_zeros();
+        for i in 0..num_neighbors {
+            let neighbor_bytes = reader.read_u8x32()?;
+            if positions & (1 << i) == 0 {
+                neighbors.push(MerkleNeighbor::Left(neighbor_bytes));
+            } else {
+                neighbors.push(MerkleNeighbor::Right(neighbor_bytes));
+            }
+        }
+        call_proof.neighbors = neighbors;
+        Ok(call_proof)
     }
 
-    #[test]
-    fn valid_disjunction2() {
-        let gens = PedersenGens::default();
+    /// Serializes the call proof to a byte array
+    pub fn encode(&self, buf: &mut Vec<u8>) {
+        encoding::write_point(self.verification_key.as_compressed(), buf);
+
+        let num_neighbors = self.neighbors.len();
+        let mut positions: u32 = 1 << 31 - num_neighbors;
```

Comment:
This length indicator seems to be growing from high to low. As number of neighbors grows, this bit is going to collide with right/left bits growing from the lowest end. Please re-read https://github.com/interstellar/slingshot/pull/268#discussion_r275136553

What we need is:

**Encoding**

Set the `1<<neighbour_count` for count indicator. Set `positions | (1<<i)` for each neighbor.

**Decoding**

Read `leading_zeros()`, if 32, fail (proof for empty tree).

Then do `neighbors_count = 31 - leading_zeros()` to determine number of neighbours (if it's 0, it means tree contains one item, there are no neighbours and no bits are used for left/right indication).

Then, use `positions & (1 << i) == 0` check to distinguish left from right for i-th neighbor.

Reply:
<img width="595" alt="image" src="https://user-images.githubusercontent.com/698/56308842-1a8dbe00-60fd-11e9-8b43-4631301b491f.png">

Reply:
Aha, I think the second comment explains my confusion! Thanks.